lib/smtp/smtp-connection.js

```diff
@@ -0,0 +1,188 @@
+        this.#configureSocket();
+        await this.send(greet());
+        await this.#handshake();
+        this.#connected = true;
+        this.emit('connect');
+      });
+    } finally {
+      this.#connectionLock.leave();
+    }
+  }
+
+  async #handshake() {
+    try {
+      this.#smtpExtensions = await this.send(ehlo(this.#name));
+      if (this.#smtpExtensions.STARTTLS && !this.#socket.encrypted) {
+        await this.send(starttls());
+        this.#upgradeConnection();
+        this.#smtpExtensions = await this.send(ehlo(this.#name));
+      }
+    } catch (err) {
+      if (err.isTerminating) throw err;
+      await this.send(helo(this.#name));
+    }
+  }
+
+  send({ type, text, successCodes, process }) {
+    if (this.#destroyed) throw new Error('SMTP connection is destroyed.');
+    if (this.#currentCommand) throw new Error(`Other command is in flight`);
+    return new Promise((resolve, reject) => {
+      this.#currentCommand = (err, code, payload) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+        if (successCodes.includes(code)) {
+          resolve(process ? process(payload) : payload);
+        } else {
+          const error = new SMTPError(code, payload, type);
+          reject(error);
+          this.emit('error', error);
+        }
+      };
+      if (text) this.#socket.write(Buffer.from(text + '\r\n', 'utf-8'));
+    });
+  }
+
+  async sendSequence(runner) {
+    await this.#sequenceLock.enter();
+    try {
+      return await runner();
+    } finally {
+      this.#sequenceLock.leave();
+    }
+  }
+
+  async sendData(dataInput) {
+    return this.sendSequence(async () => {
+      await this.send(data());
+      const promise = this.send(stream());
+      const dataStream = new SMTPDataStream();
+      dataStream.pipe(this.#socket, { end: false });
+      if (typeof dataInput.pipe === 'function') {
+        dataInput.pipe(dataStream);
+      } else {
+        dataStream.write(dataInput);
+        dataStream.end();
+      }
+      return promise;
+    });
+  }
+
+  #configureSocket() {
+    const isSecurePort = this.#port === SMTP_SECURE_PORT;
+    const connector = isSecurePort ? tls : net;
+    this.#socket = connector.connect({
+      host: this.#host,
+      port: this.#port,
+    });
+    this.#socket.once('connect', () => this.#socket.setKeepAlive(true));
+    this.#socket.on('data', (chunk) => this.#processChunk(chunk));
+    this.#socket.once('close', () => this.#destroy());
+    this.#socket.on('error', (err) => this.#destroy(err));
+  }
+
+  #upgradeConnection() {
+    const plainSocket = this.#socket;
+    plainSocket.removeAllListeners(['data']);
+    this.#socket = tls.connect({
+      host: this.#host,
+      port: this.#port,
+      socket: plainSocket,
+    });
+    this.#socket.on('data', (chunk) => this.#processChunk(chunk));
+    plainSocket.resume();
+  }
+
+  #processChunk(chunk) {
+    this.#dataBuffer += chunk;
```

Comment:
- `this.#dataBuffer.push(chunk)`
- To get collected result `Buffer.concat(this.#dataBuffer).toString()`